### PR TITLE
Speedup _contrib_index_copy

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
Re-writing the Map kernel of contrib.index_copy to speed up index_copy operator for DGL usage.
Plus a small fix to operator registration to fix problem with symbolic API.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Custom CPU kernel for best performance on CPU
- [x] Custom GPU kernel for best performance on GPU

## Comments ##
benchmark results:
CPU
- forward only (ms): 26.154723167419434->25.8462176322937  ~1x speedup
- forward+backward (ms): 63094.7536945343->79.10973119735718 ~800x speedup
GPU:
- forward only (ms): 52.20915079116821->4.262630462646484 ~12.25x speedup
- forward+backward (ms): 36445.19102573395->14.284788131713867 ~2551x speedup
```Python
import mxnet as mx

# uncomment line to use corresponding ctx
# ctx = mx.cpu()
# ctx = mx.gpu(0)

orig_row = 40000
new_row = 20000
col = 512

import random

indices = [i for i in range(orig_row)]
random.shuffle(indices)
indices = mx.nd.array(indices[0:new_row], ctx=ctx, dtype='int32')

from mxnet.test_utils import check_speed, rand_ndarray

mx_orig = rand_ndarray((orig_row, col)).as_in_context(ctx)
mx_new = rand_ndarray((new_row, col)).as_in_context(ctx)

orig = mx.sym.Variable("orig")
idx = mx.sym.Variable("idx")
new = mx.sym.Variable("new")
mx_sym = mx.sym.contrib.index_copy(old_tensor=orig, index_vector=idx, new_tensor=new)

print(check_speed(mx_sym, typ='forward', location={"orig": mx_orig, "idx": indices, "new": mx_new}, ctx=ctx, N=1000) * 1000)
print(check_speed(mx_sym, typ='whole', location={"orig": mx_orig, "idx": indices, "new": mx_new}, ctx=ctx, N=10) * 1000)
```

### CODE REVIEW

**Comment by haojin2** [comment]
@zheng-da @szha @eric-haibin-lin Please review

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Operator, pr-awaiting-review]

File: src/operator/contrib/index_copy.cc [hunk 1]
```diff
@@ -26,6 +26,99 @@
...
+                            const std::vector<OpReqType>& req,
+                            const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  CHECK_EQ(inputs.size(), 4U);
+  CHECK_EQ(outputs.size(), 3U);
+  Stream<cpu> *s = ctx.get_stream<cpu>();
+  const TBlob& out_grad = inputs[0];
+  const TBlob& index = inputs[2];
+  const TBlob& in_grad_1 = outputs[0];
```
Comment (wkcn) [inline src/operator/contrib/index_copy.cc]: Do we need assign outputs[1], namely the gradient of index to 0?

File: src/operator/contrib/index_copy.cc [hunk 2]
```diff
@@ -26,6 +26,99 @@
...
+                  const DType* out_tensor_grad,
+                  DType* orig_tensor_grad,
+                  DType* new_tensor_grad,
+                  const IType* idx,
+                  int dim_size,
+                  int idx_size) {
+    const DType* out_ptr = out_tensor_grad + i * dim_size;
+    DType* orig_ptr = orig_tensor_grad + i * dim_size;
+    std::memcpy(orig_ptr, out_ptr, sizeof(DType) * dim_size);
+    if (i < idx_size) {
```
Comment (wkcn) [inline src/operator/contrib/index_copy.cc]: It puzzled me since the meanings of variable i are different.
I recommend to use copy function to copy out_tensor_grad to orig_tensor_grad, then call the latter code.

File: src/operator/contrib/index_copy.cu [hunk 3]
```diff
@@ -18,14 +18,101 @@
...
+  template<typename DType, typename IType>
+  MSHADOW_XINLINE static void Map(int i,
+                                  const DType* out_grad,
+                                  DType* orig_grad,
+                                  DType* new_grad,
+                                  const IType* idx,
+                                  int dim_size,
+                                  int idx_size) {
+    orig_grad[i] = out_grad[i];
+    if (i / dim_size < idx_size) {
```
Comment (wkcn) [inline src/operator/contrib/index_copy.cu]: the same question as that in CPU.

**Comment by wkcn** [comment]
Great! Thanks for your contribution!
I have some comments, the rest LGTM: )

**Comment by wkcn** [comment]
The old code supports OpReq, but it is cancelled in this PR.

File: src/operator/contrib/index_copy.cc (hunk 2 above)
Comment (wkcn) [reply src/operator/contrib/index_copy.cc]: There is a test case: index[0]=3，i=0,1,2,3
orig_tenser_grad[3] will be rewritten when i==3

File: src/operator/contrib/index_copy.cc (hunk 1 above)
Comment (haojin2) [reply src/operator/contrib/index_copy.cc]: indices should have no gradients.

File: src/operator/contrib/index_copy.cc (hunk 2 above)
Comment (haojin2) [reply src/operator/contrib/index_copy.cc]: okay

**Comment by haojin2** [comment]
@wkcn I don't think ReqType really matters here, cause kAddTo for grad would hardly make any sense...

**Comment by wkcn** [comment]
So I think it is better to check whether OpReq is kWrite or kInplace.

**Comment by szha** [comment]
What if this op is used during training where gradient accumulation is used?

**Comment by haojin2** [comment]
@wkcn kWriteTo and kWriteInplace actually share the same code in KERNEL_ASSIGN: https://github.com/apache/incubator-mxnet/blob/master/src/operator/mxnet_op.h#L304-L319 so no need to distinguish between them, I'll add the kAddTo support as per @szha's request.

**Comment by wkcn** [comment]
@haojin2 Sorry that I forgot it. After adding the kAddTo, LGTM : )

**Comment by wkcn** [comment]
I think we should CHECK(req[0]==kWriteTo||req[0]==kInplace); in the forward function.

File: src/operator/contrib/index_copy.cc [hunk 4]
```diff
@@ -26,6 +26,121 @@
...
+                           const OpContext& ctx,
+                           const std::vector<TBlob>& inputs,
+                           const std::vector<OpReqType>& req,
+                           const std::vector<TBlob>& outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  CHECK_EQ(inputs.size(), 3U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  mshadow::Stream<cpu> *s = ctx.get_stream<cpu>();
```
[the remark this answers is not in the document]
Comment (haojin2) [reply src/operator/contrib/index_copy.cc]: done

**Comment by haojin2** [comment]
@szha @zheng-da Ready for merge I think.
